odoo/addons/base/models/ir_actions.py

```diff
@@ -564,17 +564,17 @@ def _run_action_multi(self, eval_context=None):
             res = act.run() or res
         return res
 
-    def _run_action_object_write(self, eval_context=None):
-        """Apply specified write changes to active_id."""
+    def _run_action_object_write_multi(self, eval_context=None):
+        """Apply specified write changes to active_ids."""
         vals = self.fields_lines.eval_value(eval_context=eval_context)
         res = {line.col1.name: vals[line.id] for line in self.fields_lines}
 
-        if self._context.get('onchange_self'):
+        if len(self) == 1 and self._context.get('onchange_self'):
             record_cached = self._context['onchange_self']
             for field, new_value in res.items():
                 record_cached[field] = new_value
         else:
-            self.env[self.model_id.model].browse(self._context.get('active_id')).write(res)
+            self.env[self.model_id.model].browse(self._context.get('active_ids', self._context.get('active_id', []))).write(res)
```

Comment:
For the sake of readability, it might be worth splitting this line...
```suggestion
            record_ids = self._context.get('active_ids', self._context.get('active_id', ()))
            records = self.env[self.model_id.model].browse(record_ids)
            records.write(res)
```